### ISSUE TITLE
Dependency patch resolution: Only allow patches from dependencies you trust

### ISSUE THREAD

**Issue by RobinHoutevelts** [cweagans/composer-patches#551]
### Verification

- [X] I have updated Composer to the most recent stable release (`composer self-update`)
- [X] I have updated Composer Patches to the most recent stable release (`composer update cweagans/composer-patches`)
- [X] I am using one of the supported PHP versions (8.0+)
- [X] I have searched existing issues _and_ discussions for my idea.

### Is your feature request related to a problem?

I only allow patches from sources I trust.
By default `cweagans/composer-patches` seems to apply patches from any of my dependencies.

### Describe your proposed solution(s)

```json
{
   // don't allow patches from anyone else
   "ignore-dependency-patches": ["*"],
   // list of repositories you trust to apply patches from
   "dependency-patches": ["my-org/drupal-patches"]
}
```

### Additional context

You will have less bug reports related to Dependency patch resolution if users are forced to know where they want to apply patches from.

However, I don't mind it -not my project- but I would need a way to opt-out and instead define the sources I trust myself.

**Comment by cweagans** [comment]
I'm on board. Maybe we can add another config value 'allow-dependency-patches' and have that act as an allow list if set. I don't want to have to set both. Having both set should probably be an error.

**Comment by RobinHoutevelts** [comment]
Thanks!

> Having both set should probably be an error.

Yes, that would also make this much better to maintain 👍

**Comment by RobinHoutevelts** [comment]
If I want to disable dependency patch resolution altogether: 
Would setting `allow-dependency-patches: []` (empty list) achieve that?

**Comment by cweagans** [comment]
You can just disable the resolver - no need to try to make an allow list work for that :)

**Comment by RobinHoutevelts** [comment]
@cweagans I'd be happy to take a stab at adding `allow-dependency-patches`.
But I'm currently busy for the next two weeks.

**Comment by cweagans** [comment]
No worries. Same thing here - I'll update here if I beat you to it!

**Comment by cweagans** [comment]
Correct. If the option is NOT defined, all dependencies will be scanned for patches. If it IS defined, only the listed packages will be scanned.

**Comment by Natshah** [comment]
I'm so interested to having this option, something like allow-plugins but it could be called `allowed-patches` or `allowed-patchers` 
as we do have `allowed-packages` for `drupal-scaffold`
```
      "allowed-packages": [
        "drupal/core", // Only listing to demo
        "vardot/varbase"
      ],
```
as
```
      "allow-dependency-patches": [
        "drupal/core", // Only listing to demo
        "vardot/varbase-patches"
      ]
```
Maybe in a project they could select to add more 

```
      "allow-dependency-patches": [
        "vardot/varbase-patches",
        "my-trusted-privet-vendor/private-list-of-patches"
      ]
```

**Comment by cweagans** [comment]
@Natshah All composer patches config is nested inside of a `extra.composer-patches` to avoid naming conflicts like that.

As an aside, you might be interested in https://github.com/cweagans/composer-configurable-plugin/ as a replacement for this file: https://github.com/drupal/core-composer-scaffold/blob/11.x/ScaffoldOptions.php

It handles getting config from composer.json + the environment, merging it all down into one location, and ensures that it's the right data type. It probably wouldn't be backwards compatible, but it might be an interesting thing to explore!

**Comment by bradjones1** [comment]
Another use case here, again from Drupal-land. Now that Drupal CI is run on GitLab the template allows modules to express patches in their `composer.json` which will be applied during CI runs, e.g. if they require a patch to core. I'm in a situation where the patches in question conflict with others I have on my specific project, so when they are applied from the dependency they fail. Would be nice to be able to disable the patches to avoid this conflict.

**Comment by metalbote** [comment]
I think this is a very important feature, especially when i think about security, implemented this and PR

**Comment by regilero** [comment]
While thinking about securely supporting patches, I'd like for example to only allow patches added in a local folder, and present in the first level of my composer.json.
I'm not sure what should be done of untrusted patches referenced by packages, is an error message enough? Should we make the build fail?
But allowing any package of patching the other ones seems strange.
